### PR TITLE
feat(adapters): C-104 — Discord+Mattermost SurfaceAdapter face (MIG-3b)

### PR DESCRIPTION
**Pull request by mellanon**
## What

MIG-3.4 + 3.5 + 3.6: each adapter now exposes a `surfaceConfig: SurfaceAdapter` face so the surface-router (cortex#25/#26/#27) can fan bus envelopes to them. Adapters still own their platform-side connection (Discord gateway, Mattermost websocket); they no longer need to think about NATS subscriptions — that's the runtime's job.

## Files

- **`src/adapters/envelope-renderer.ts`** (new, ~30 LOC) — shared `formatEnvelopeAsMarkdown(envelope)` pure function. Both Discord and Mattermost render markdown identically.
- **`src/adapters/mock.ts`** — adds `envelopesRendered`, `surfaceSubjects`, `surfaceConfig` getter
- **`src/adapters/discord/index.ts`** — adds `DiscordAdapterConfig.{surfaceSubjects, surfaceFilter, surfaceFallbackChannelId}`, `surfaceConfig` getter, `renderEnvelope` private. Uses existing `postResponse` so the disconnected-Discord pending-result buffer is reused.
- **`src/adapters/mattermost/index.ts`** — same shape. Uses `postReply` directly; JetStream replay covers redelivery (per design §3.3).

## SurfaceAdapter contract

```typescript
const router = createSurfaceRouter(runtime);
router.register(discordAdapter.surfaceConfig);  // ← new pattern
await router.start();
// envelopes flow runtime → router → adapter.render() → platform
```

## Render() implementation (v1)

`renderEnvelope(envelope)` posts a markdown code-block of the envelope to the adapter's configured fallback channel. Two log+drop failure modes:
1. Client not ready (Discord) — log warn, drop
2. No fallback configured — log warn, drop

Never throws (matches `renderWithIsolation`'s contract assumption in surface-router).

Per-event-type templates + payload-driven channel routing land at the **Renderer model** in MIG-7.2d. v1 is intentionally simple.

## Tests added (22 new)

- **`src/adapters/__tests__/surface-integration.test.ts`** (6 tests): router-register-mock-dispatch path, end-to-end runtime→handler→router→adapter, subject-mismatch no-op, two adapters with different subjects, default `mock.>` regression pin, §3.5 acknowledgement (adapters don't open NATS subs).
- **`src/adapters/discord/__tests__/render-envelope.test.ts`** (16 tests): surfaceConfig getter shape (id/subjects/filter/render-binding), happy path (channel id, type header, correlation_id present/absent, JSON code block), failure modes (client not ready, client null, no fallback channel, never throws).

## Verification

- cortex root \`bunx tsc --noEmit\` → **exit 0**
- \`bun test src/adapters/\` → **81/81 pass** (was 60; **+21 new** including the 22 above; one was a fixture rewrite)
- Full cortex suite — **1667 pass / 1 pre-existing fail** (the env-dependent React shell test from MIG-2)

## Out of scope (explicit deferrals to MIG-3b-ii)

- **3.7** `system.adapter.{disconnected,degraded,recovered}` envelope publishing — needs a `runtime.publish()` surface that doesn't exist yet (today's MyelinRuntime is receive-only)
- **3.8** `system.dispatch.aborted` envelope publishing — same dependency

Both land in **MIG-3b-ii** after a small `runtime.publish()` PR lands the publish surface.

## Plan grounding

`docs/architecture.md` §8 (adapters/) + §9 (Renderer model). Plan §4 MIG-3.4/3.5/3.6. G-1111 spec §4.1 SurfaceAdapter interface (implemented; lives in `src/bus/surface-router.ts`).

Refs cortex#5 (MIG-3 umbrella).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/adapters/envelope-renderer.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ *   **{type}** [correlation_id?]
+ *   ```json
+ *   {payload}
+ *   ```
+ *
+ * The correlation_id is included when present so an operator scanning
+ * the channel can correlate envelopes across a workflow without opening
+ * the envelope details.
+ */
+export function formatEnvelopeAsMarkdown(envelope: Envelope): string {
```
Comment (mellanon) [inline src/adapters/envelope-renderer.ts:32]: **[suggestion/quality]** No direct unit tests for `formatEnvelopeAsMarkdown`.

The helper is a 9-line pure function and is exercised end-to-end by the Discord happy-path tests (`render-envelope.test.ts`: "contains envelope.type as bold header", "contains correlation_id when present", "omits correlation bracket when absent", "contains payload as JSON code block"). Coverage is real, just routed through the adapter.

Suggestion: add a tiny `__tests__/envelope-renderer.test.ts` next to the helper covering edge cases (`payload: {}`, `payload: null`, deeply nested payloads, payloads with non-JSON-safe values like circular refs). Not blocking — current coverage is sufficient for v1 — but a 6-test file would future-proof the helper as the v2 Renderer model lands and this becomes the fallback path.

File: src/adapters/mattermost/index.ts [hunk 2]
```diff
@@ -275,4 +288,63 @@ export class MattermostAdapter implements PlatformAdapter {
...
+   * v1: post the formatted envelope to the configured fallback channel via
+   * `postReply` (no rootId — top-level post in the channel). v2 routing
+   * lives in the Renderer model (MIG-7.2d).
+   *
+   * Failure mode: `postReply` may throw on HTTP error; we catch + log
+   * here so the surface-router's `renderWithIsolation` doesn't have to
+   * be the only safety net. Dropping is acceptable because JetStream
+   * replay handles redelivery.
+   */
+  private async renderEnvelope(envelope: Envelope): Promise<void> {
```
Comment (mellanon) [inline src/adapters/mattermost/index.ts:336]: **[warning/quality]** No `MattermostAdapter` unit tests for `renderEnvelope`.

The integration test (`surface-integration.test.ts`) uses MockAdapter to assert the dispatch path, and DiscordAdapter has 16 unit tests for `renderEnvelope` covering the surfaceConfig shape + happy path + failure modes. Mattermost has the same surface contract but zero unit tests — the "no fallback channel" log+drop path, the `postReply` try/catch, and the surfaceConfig getter shape are all unverified.

The adapters are symmetric in shape but asymmetric in coverage. Suggest mirroring `discord/__tests__/render-envelope.test.ts` for Mattermost: same 16-test structure, swap the fake `client.channels.fetch` for a fake `fetch` that intercepts the `postReply` HTTP call. Even a 6-test minimum (getter shape + happy path + no-fallback-channel + postReply-throws) would close the asymmetry.

Not strictly blocking for v1 (the integration test proves the dispatch path; JetStream replay covers the recovery path per design-cortex.md §3.3), but it's the kind of gap that bites later when someone refactors the renderer model in MIG-7.2d.

File: src/adapters/discord/index.ts [hunk 3]
```diff
@@ -608,4 +623,75 @@ export class DiscordAdapter implements PlatformAdapter {
...
+   *
+   * Wiring: the bot's startup composer calls `router.register(adapter.surfaceConfig)`
+   * once per adapter, after MyelinRuntime has started. The router never opens
+   * a NATS subscription — that's MyelinRuntime's job (per spec §5.1) — so this
+   * face is purely the rendering hook.
+   */
+  get surfaceConfig(): SurfaceAdapter {
+    return {
+      id: this.instanceId,
+      subjects: this.adapterConfig.surfaceSubjects ?? [],
```
Comment (mellanon) [inline src/adapters/discord/index.ts:661]: **[suggestion/architecture]** Empty `surfaceSubjects` makes the adapter silently inert.

`subjects: this.adapterConfig.surfaceSubjects ?? []` means an adapter registered with no `surfaceSubjects` configured renders nothing — the surface-router's `subjectMatches` loop iterates zero times, returns false, adapter never sees an envelope. No error, no warn.

This is correct behaviour (a bot deployed before its YAML adds `surfaceSubjects` shouldn't crash), but it's a silent footgun. Two options worth considering for MIG-3b-ii:

1. Log a one-time warn at first dispatch attempt: "discord-{id} registered with empty surfaceSubjects; envelopes will not render until subjects are configured."
2. Move the empty-subjects check into a startup-time validator that warns once at composer time.

Today this is fine — getting the contract shape right matters more than the polish. Worth a TODO comment so it's not forgotten.

File: src/adapters/discord/index.ts [hunk 4]
```diff
@@ -608,4 +623,75 @@ export class DiscordAdapter implements PlatformAdapter {
...
+   * routing based on `envelope.payload` (e.g. `payload.repo` → repo-specific
+   * channel) and sovereignty-aware redaction. This method stays as the
+   * default fallback for envelopes that don't match a registered template.
+   *
+   * Failure modes: this method never throws — `postResponse` already swallows
+   * delivery errors and buffers when disconnected. The router's
+   * `renderWithIsolation` wraps us in a timeout regardless.
+   */
+  private async renderEnvelope(envelope: Envelope): Promise<void> {
+    if (!this.client?.isReady()) {
```
Comment (mellanon) [inline src/adapters/discord/index.ts]: **[nit/quality]** `client.isReady()` covers null-client + not-ready-client uniformly.

`if (!this.client?.isReady())` — the optional chain returns `undefined` when `this.client` is null, and `!undefined` is `true`, so this single check correctly handles both pre-start (`client === null`) and disconnected (`client !== null && !isReady()`).

The two unit tests ("drops + warns when client is not ready" and "drops + warns when client is null") both end up at the same `console.warn("client not ready")` branch. Tests are fine as-is, but the warn message could distinguish the two states ("client null — adapter not started" vs "client present but not ready — disconnected") to aid debugging when someone is staring at logs at 2am.

Pure nit — current behaviour is correct.

**Comment by mellanon** [comment]
## Full Review Summary: cortex#28

Comprehensive review applying all 6 lenses + duplication. PR delivers MIG-3.4/3.5/3.6 — DiscordAdapter, MattermostAdapter, and MockAdapter all expose a `surfaceConfig: SurfaceAdapter` getter + `renderEnvelope()` method matching the §4.1 contract from `g-1111-spec.md`.

### Lens Results

| Lens | Critical | Warning | Suggestion | Nit | Verdict |
|------|----------|---------|------------|-----|---------|
| CodeQuality | 0 | 1 | 1 | 1 | advisory |
| Security | 0 | 0 | 0 | 0 | pass |
| Hardening | 0 | 0 | 0 | 0 | pass |
| Architecture | 0 | 0 | 1 | 0 | advisory |
| EcosystemCompliance | 0 | 0 | 0 | 0 | pass |
| Performance | 0 | 0 | 0 | 0 | pass |
| Duplication | 0 | 0 | 0 | 0 | pass |

### Contract verification

- **`surfaceConfig` shape** — All three adapters return `{id, subjects, filter?, render}`. `health?` correctly omitted (reserved for G-1111.E per surface-router.ts:59).
- **`render` binding** — Discord/Mattermost use arrow form `(envelope) => this.renderEnvelope(envelope)`; Mock uses inline arrow. The "render is bound to the adapter" test (render-envelope.test.ts:181) pulls the function off the getter and invokes it standalone — proves `this` survives.
- **`render()` never throws** — DiscordAdapter routes through `postResponse` which already swallows + buffers via the existing pending-result mechanism. MattermostAdapter wraps `postReply` in try/catch + log. MockAdapter is a pure push to `envelopesRendered`. Combined with surface-router's `renderWithIsolation`, the §5.3 isolation guarantee holds.
- **`PayloadFilter` import path** — `import type { PayloadFilter } from "../../bus/payload-filter";` resolves correctly in both adapters; tsc passes clean.
- **Empty `surfaceSubjects` behaviour** — Returns `[]`, surface-router's `subjectMatches` iterates zero times, adapter never renders. Correct fail-safe (no crash) but silent — see inline finding.

### Test coverage

| Suite | Tests | New | Notes |
|-------|-------|-----|-------|
| `src/adapters/__tests__/surface-integration.test.ts` | 6 | +6 | Proves register → dispatch → render path; covers subject mismatch + multi-adapter routing + §3.5 (adapters don't open NATS subs) |
| `src/adapters/discord/__tests__/render-envelope.test.ts` | 16 | +16 | surfaceConfig getter shape + happy path + 4 failure modes |
| `src/adapters/mattermost/__tests__/` | — | — | **Missing** — no unit tests for MattermostAdapter renderEnvelope |
| **Total adapters/** | **81/81** | **+21** | (one was a fixture rewrite; PR reports +22 because of the rewrite delta) |

### Verification (from worktree)

- `bunx tsc --noEmit` → exit 0 ✓
- `bun test src/adapters/` → 81 pass / 0 fail ✓
- Full cortex suite → 1667 pass / 1 pre-existing fail (env-dependent React shell test from MIG-2; unrelated) ✓

### Inline findings (4)

1. [warning/quality] `src/adapters/mattermost/index.ts:327` — No MattermostAdapter unit tests for renderEnvelope; the surfaceConfig shape, no-fallback-channel drop, and postReply try/catch are unverified at the unit level. Integration test covers dispatch path, but adapter-specific failure modes are gaps.
2. [suggestion/architecture] `src/adapters/discord/index.ts:650` — Empty `surfaceSubjects` makes the adapter silently inert. Correct fail-safe, but log-once would help operators catch missing config.
3. [suggestion/quality] `src/adapters/envelope-renderer.ts:32` — Pure helper has no direct tests (covered transitively via Discord). Adding 6 edge-case tests would future-proof for v2 Renderer model (MIG-7.2d).
4. [nit/quality] `src/adapters/discord/index.ts:675` — Single warn message for both null-client and not-ready-client; distinguishing them aids 2am debugging.

### Architectural observation: getter vs. `toSurfaceAdapter()`

The `get surfaceConfig(): SurfaceAdapter` form is the right choice here. It signals "read-only view derived from current adapter state" and matches the spec's *"the adapter interface the surface-router consumes"* framing. A `toSurfaceAdapter()` method would imply transformation cost or side effects that don't exist. The getter creates a fresh object per access, but `register()` returns an unregister handle that captures the adapter by reference (not by identity), so identity-equality concerns don't apply to the consumer API. Approve the design.

### Overall Verdict: request-changes (advisory)

The contract is faithful to §4.1, render-never-throws holds across all three adapters, and the integration test proves dispatch end-to-end. The Mattermost test gap (warning) and three suggestions/nits don't block correctness — they're polish for the v2 hand-off. v1 is functionally sound; mergeable after the Mattermost test gap is addressed (or explicitly deferred to MIG-3b-ii with a tracking comment).

**Review by mellanon — COMMENTED** [review]
Full review (6 lenses + duplication) — SurfaceAdapter contract holds across all 3 adapters; render-never-throws verified; dispatch path proven by integration tests; one warning (no MattermostAdapter unit tests for renderEnvelope) + 2 suggestions + 1 nit.

verdict: blockers=0 majors=1 nits=3 — recommend: request-changes (posted as comment-review — bot account opened the PR; --request-changes blocked by GitHub)

**Review by mellanon — COMMENTED** [review]
## Round-2 Review (Echo)

Verified PR head `a3fffca` against the four round-1 items. All resolved cleanly.

### Round-1 verification

- **W (Mattermost test gap) — RESOLVED.** New `src/adapters/mattermost/__tests__/render-envelope.test.ts` (316 LOC, **20 tests**) mirrors the Discord shape: surfaceConfig getter (6), happy path (7), failure modes (4), empty-subjects warn (3). Asserts are non-trivial — `globalThis.fetch` is intercepted and the parsed POST body checks `channel_id`, absence of `root_id` (top-level, not threaded), bold-`**type**` header, correlation bracket presence/absence, JSON code block, and `Authorization: Bearer test-token`. Symmetry with Discord is now real.

- **S1 (envelope-renderer tests) — RESOLVED.** New `src/adapters/__tests__/envelope-renderer.test.ts` (151 LOC, **11 tests**) pins the v1 fallback formatter contract directly. Highlights: line-by-line shape pinned for the canonical case (lines[0..5] each asserted), JSON.parse round-trip on a nested payload to prove no field drops, regression guard against trailing-space header when `correlation_id` absent, exact `\`\`\`json\n{}\n\`\`\`` shape for empty payload, and explicit assertion that output never ends in `\n`. Good defence-in-depth before MIG-7.2d.

- **S2 (empty-subjects warn) — RESOLVED.** Both constructors guard with `if (adapterConfig.surfaceSubjects?.length === 0)` (`discord/index.ts:80`, `mattermost/index.ts:77`). Optional chaining is the right primitive: `undefined?.length === 0` is `false` → silent on opt-out; `[]?.length === 0` is `true` → warn. Six tests across the two adapter test files cover all three branches (`[]` warns, `undefined` silent, populated silent). Warn copy is identical between adapters.

- **N1 (null vs not-ready warns) — RESOLVED.** `DiscordAdapter.renderEnvelope` (`discord/index.ts:694–705`) splits the guard cleanly: `this.client === null` → "called before start()" (config bug); `!this.client.isReady()` → "called while shard reconnecting" (transient). Tests at `discord/__tests__/render-envelope.test.ts:256–278` assert each branch positively AND assert the *other* message is absent — proves the two paths cannot collapse back together.

### Verification

- `bunx tsc --noEmit` → exit 0 (clean)
- `bun test src/adapters/` → **115 pass / 0 fail**, 242 expects (was 81 → +34, matches the round-1 deltas)
- `bun test` (full) → **1701 pass / 1 fail**. The single failure is `src/surface/mc/__tests__/server.test.ts > serves the React dashboard shell on GET /` — ENOENT on `src/dist/dashboard-v2/index.html`, a missing build artefact. Last-touched by C-103 (#14), unrelated to this PR's diff (adapter-files-only).
- CI green (Compass governance validators ✓), `mergeable: MERGEABLE`, `mergeStateStatus: CLEAN`.

### New findings

None.

### Verdict

**APPROVE.** All four round-1 items resolved with appropriately tight tests; no regressions; no new findings. Ship it.

**Comment by mellanon** [comment]
Round-2: **APPROVE** (posted as comment-review — bot-self constraint: my GH account opened the PR, --approve blocked).

All 4 round-1 items resolved at `a3fffca`:
- W (Mattermost tests): +20 tests, fetch-intercepted, mirrors Discord shape
- S1 (envelope-renderer): +11 pure-fn tests, line-by-line shape pinned, JSON round-trip
- S2 (empty-subjects warn): `?.length === 0` in both constructors, 6 tests across 3 branches
- N1 (null vs not-ready): split into two distinct warns, tests assert each branch positively + the other absent

`bunx tsc --noEmit` clean | `bun test src/adapters/` 115/115 (+34) | full `bun test` 1701 pass + 1 unrelated pre-existing fail (dashboard shell ENOENT, last-touched by C-103). CI green, mergeable. Ship it.
